### PR TITLE
scheduler: fix race condition in equivalence cache

### LINKED ISSUE #62921: Race condition with scheduler equivalence cache invalidation
/kind bug
/sig scheduling

**What happened**:
In https://github.com/kubernetes/kubernetes/issues/61512 it was discovered that certain event sequences cause stale, incorrect entries to remain in the equivalence cache. Quoting https://github.com/kubernetes/kubernetes/issues/61512#issuecomment-375830364,

> Here is what happens:
> 
> 1. Scheduler tries to schedule Pod A, but A is not schedulabe and requires preemption.
> 1. Scheduler preempts several Pods to make room for Pod A.
> 1. Pod deletion events start arriving for the preempted pods.
> 1. Deletion events trigger scheduling queue and scheduler tries scheduling Pod A again.
> **Here are the important events (the race is between step 6 and 7):**
> 1. Scheduler determines that Pod A is not yet schedulable, because deletion of the last preempted Pod is not reflected in the scheduler cache. Let's call the last preempted Pod X.
> 1. Deletion of Pod X is observed and invalidates equivalence cache.
> 1. Scheduler updates equivalence cache by the result of step 5.
> 1. Since scheduler observed deletion of Pod X while it was attempting to schedule A, scheduler puts A back in the active queue to try again.
> 1. Scheduler tries again, but it uses equivalence cache results this time. The results which was put in step 7. Scheduler determines that A is not schedulable based on equivalence cache and no other event invalidates the equivalence cache.
> Test times out.
> 

**How to reproduce it (as minimally and precisely as possible)**:
Cherry pick https://github.com/kubernetes/kubernetes/pull/58555 (which [has](https://github.com/kubernetes/kubernetes/pull/58903) [been](https://github.com/kubernetes/kubernetes/pull/60753) [reverted](https://github.com/kubernetes/kubernetes/pull/61621)) and TestPreemptionStarvation will become flaky.

I am currently writing a test to reproduce this deterministically.

/assign misterikkit

### PR DESCRIPTION
**Pull request by misterikkit**
**What this PR does / why we need it**:
This adds an equivalence cache test to exercise the race condition observed in https://github.com/kubernetes/kubernetes/issues/62921 and then fixes the race.

The `Cache` interface needed a new method to check whether a `NodeInfo` is stale, and `genericScheduler` needed some plumbing to make the `Cache` object available to `podFitsOnNode()`.

The solution is, right before writing to the eCache, check the scheduler cache to see if the current `NodeInfo` object is out of date. If the node is out of date, then don't write to the eCache. If the `NodeInfo` is stale, it is because of a cache update that should also invalidate the eCache entry. That invalidation either happens before `podFitsOnNode()` acquires the eCache lock (original bug, so we don't do the write) or blocks until we release that lock (removing the potentially bad entry).

Fixes #62921 

**Special notes for your reviewer**:

**Release note**:

equivalence cache is still alpha, so no release note.
```release-note
NONE
```
/sig scheduling
/assign bsalalamat
/assign resouer

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -547,10 +539,13 @@ func podFitsOnNode(
 							} else {
 								predicateResults[predicateKey] = HostPredicate{Fit: fit, FailReasons: reasons}
 							}
-							result := predicateResults[predicateKey]
-							ecache.UpdateCachedPredicateItem(
-								pod.GetName(), info.Node().GetName(),
-								predicateKey, result.Fit, result.FailReasons, equivCacheInfo.hash, false)
+							// Skip update if NodeInfo is stale.
+							if cache != nil && cache.IsUpToDate(info) {
```
Comment (resouer) [inline pkg/scheduler/core/generic_scheduler.go]: Thanks! Just took a first round of review, I think this actually consistent with this comment: https://github.com/kubernetes/kubernetes/issues/61512#issuecomment-377577199, which in my opinion, should solve the problem. 

I need to check more edge cases tomorrow.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 2]
```diff
@@ -649,3 +655,84 @@ func TestInvalidateAllCachedPredicateItemOfNode(t *testing.T) {
...
+func (c *syncingMockCache) UpdateNodeNameToInfoMap(infoMap map[string]*schedulercache.NodeInfo) error {
+	err := c.Cache.UpdateNodeNameToInfoMap(infoMap)
+	c.once.Do(func() {
+		c.cycleStart <- struct{}{}
+		<-c.cacheInvalidated
+	})
+	return err
+}
+
+// Test that equivalence cache invalidation is correctly handled when an invalidation event happens early in a scheduling cycle. Specifically, the event occurs after schedulercache is snapshotted and before equivalence cache lock is acquired.
```
Comment (bsalamat) [inline pkg/scheduler/core/equivalence_cache_test.go]: nit: please start the comment with the function name.

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 3]
```diff
@@ -649,3 +655,84 @@ func TestInvalidateAllCachedPredicateItemOfNode(t *testing.T) {
...
+	eCache := NewEquivalenceCache(fakeGetEquivalencePod)
+	// Ensure that equivalence cache invalidation happens after the scheduling cycle starts, but before the equivalence cache would be updated.
+	go func() {
+		<-mockCache.cycleStart
+		pod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "new-pod", UID: "new-pod"}, Spec: v1.PodSpec{NodeName: "machine1"}}
+		if err := cache.AddPod(pod); err != nil {
+			t.Errorf("Could not add pod to cache: %v", err)
+		}
+		eCache.InvalidateAllCachedPredicateItemOfNode("machine1")
+		mockCache.cacheInvalidated <- struct{}{}
```
Comment (bsalamat) [inline pkg/scheduler/core/equivalence_cache_test.go]: Kudos for writing this sophisticated test to verify that the new synchronization mechanism works!

**Review by bsalamat — COMMENTED** [review]
Just a minor nit. Otherwise, LGTM. I wait for Harry to finish his review before approving.
You can squash commits now.

File: pkg/scheduler/core/equivalence_cache_test.go (hunk 2 above)
Comment (misterikkit) [reply pkg/scheduler/core/equivalence_cache_test.go]: done

**Comment by misterikkit** [comment]
I resolved some build errors after rebasing. Hopefully ci-bot is happy after this.

**Comment by justinsb** [comment]
/test pull-kubernetes-e2e-kops-aws

Sorry - I think I manually nuked your DNS names mid-test

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (bsalamat) [reply pkg/scheduler/core/generic_scheduler.go]: Yes. Luckily we already had a generation number which seemed reasonable to use.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 4]
```diff
@@ -465,18 +470,21 @@ func TestFindFitAllError(t *testing.T) {
...
+		cache.AddNode(&v1.Node{ObjectMeta: metav1.ObjectMeta{Name: n}})
+	}
 	predicates := map[string]algorithm.FitPredicate{"true": truePredicate, "matches": matchesPredicate}
+	prioritizers := []algorithm.PriorityConfig{{Map: EqualPriorityMap, Weight: 1}}
+
+	s := NewGenericScheduler(
+		cache, nil, NewSchedulingQueue(), predicates, algorithm.EmptyPredicateMetadataProducer, prioritizers, algorithm.EmptyPriorityMetadataProducer, nil, nil, nil, false, false)
+	scheduler := s.(*genericScheduler)
+	cache.UpdateNodeNameToInfoMap(scheduler.cachedNodeInfoMap)
+
```
[the remark this answers is not in the document]
Comment (misterikkit) [reply pkg/scheduler/core/generic_scheduler_test.go]: Done

File: pkg/scheduler/core/equivalence_cache_test.go [hunk 5]
```diff
@@ -649,3 +655,84 @@ func TestInvalidateAllCachedPredicateItemOfNode(t *testing.T) {
...
+func (c *syncingMockCache) UpdateNodeNameToInfoMap(infoMap map[string]*schedulercache.NodeInfo) error {
+	err := c.Cache.UpdateNodeNameToInfoMap(infoMap)
+	c.once.Do(func() {
+		c.cycleStart <- struct{}{}
+		<-c.cacheInvalidated
+	})
+	return err
+}
+
+// TestEquivalenceCacheInvalidationRace tests that equivalence cache invalidation is correctly handled when an invalidation event happens early in a scheduling cycle. Specifically, the event occurs after schedulercache is snapshotted and before equivalence cache lock is acquired.
```
Comment (resouer) [inline pkg/scheduler/core/equivalence_cache_test.go]: These newly added tests have several super long comments and lines, mind to wrap them?

**Review by resouer — APPROVED** [review]
Already slacked with Jonathan, I will LGTM except some minor nits, thanks a lot for taking this up!

**Review by bsalamat — COMMENTED** [review]
/lgtm

**Comment by misterikkit** [comment]
> @misterikkit Are you planning to do the revert * 4 PR?
Sure, I will create that PR later today.

**Comment by resouer** [comment]
Related: https://github.com/kubernetes/kubernetes/pull/63264

**Pull request merged**